File: Data Structures & Algorithms/word-break/submission-1.cpp
class Solution {
public:
    bool dp(const unordered_set<string>& dict, string& s, vector<int>& memo, int i) {
        if (i == s.length()) {
            return true;
        }

        if (memo[i] >= 0) {
            return memo[i];
        }

        for (int j = 0; i+j <= s.length(); j++) {
            string sstr = s.substr(i, j);
            cout << sstr << '\n';
            if (find(dict.begin(), dict.end(), sstr) != dict.end()) {
                if (dp(dict, s, memo, i+j)) {
                    memo[i] = 1;
                    return true;
                }
            }
        }

        memo[i] = 0;
        return false;
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict;

        for (string& word : wordDict) {
            dict.insert(word);
        }

        vector<int> memo(s.length(), -1);
        bool res = dp(dict, s, memo, 0);
        for (bool b : memo) {
            cout << b << ' ';
        }
        return res;
    }
};

File: Data Structures & Algorithms/pow-x-n/submission-1.py
class Solution:
    def recursion(self, x, n):
        if n == 0:
            return 1

        if n % 2 == 0:
            return self.recursion(x*x, n//2)
        else:
            return x * self.recursion(x*x, (n-1)//2)

    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1

        is_neg = False
        if n < 0:
            n = -n
            is_neg = True
        
        res = self.recursion(x, n)

        if is_neg:
            res = 1/res

        return res